### PR TITLE
storage: unsafe key mangling for MVCCIterator

### PR DESCRIPTION
**Pull request by sumeerbhola**
It can uncover bugs in code that misuses unsafe
keys. It uncovered the bug fixed in
https://github.com/cockroachdb/cockroach/pull/60460/files#diff-84108c53fd1e766ef8802b87b394981d3497d87c40d86e084f2ed77bba0ca71a

Release note: None

### CODE REVIEW

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60499)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/storage/intent_interleaving_iter.go, line 902 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60499#-MT_r-NJ06UnBpibYB2v:-MT_r-NJ06UnBpibYB2w:bsvqb8c) ([raw file](https://github.com/cockroachdb/cockroach/blob/797aeba5eb3e8e540816d43349b77eb11a70cae2/pkg/storage/intent_interleaving_iter.go#L902)):*
> ```Go
> 	if rand.Intn(2) == 0 {
> 		for _, b := range [3][]byte{i.keyBuf, i.rawKeyBuf, i.rawMVCCKeyBuf} {
> 			// for _, b := range [3][]byte{i.keyBuf, i.rawKeyBuf, i.rawMVCCKeyBuf} {
> ```

don't forget to remove this before it merges

**Comment by sumeerbhola** [comment]
I forgot to mark this Do Not Merge. I've done that now. I had not added reviewers since this was an experiment for which I needed a full set of CI test runs -- is there is a simpler way to do that without creating a PR?

**Comment by knz** [comment]
Creating a PR is fine but github gives you the option to mark it as "draft" now so perhaps you could use that.

**Review by sumeerbhola — COMMENTED** [review]
This is ready for a look. Reviewable is confused because of the history of this PR and shows too many files -- if one selects "combine commits for review", it will look sane.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60499#-:-MTqPOmYCo3N0e1o09IX:b-bfcxs3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @nvanbenschoten, @petermattis, and @sumeerbhola)

**Review by nvb — APPROVED** [review]
:lgtm: cool idea! Does this have a noticeable impact on racetest duration?

> Reviewed 1 of 4 files at r3, 3 of 18 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60499#-:-MTqZlC92xd99TJ1Bcvq:b9e55yd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal, @petermattis, and @sumeerbhola)

**Review by sumeerbhola — COMMENTED** [review]
TFTR!

> Does this have a noticeable impact on racetest duration?

The racetests run for this PR look fast enough, with the slowest taking 7s
https://teamcity.cockroachdb.com/viewLog.html?buildId=2691122&buildTypeId=Cockroach_UnitTests_Testrace&tab=testsInfo

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60499#-:-MTqcEvcDijjzORnjmja:b-e127o9)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal, @knz, and @petermattis)

___
*[pkg/storage/intent_interleaving_iter.go, line 902 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60499#-MT_r-NJ06UnBpibYB2v:-MTqc0MFD4lZF-LGtyUZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/797aeba5eb3e8e540816d43349b77eb11a70cae2/pkg/storage/intent_interleaving_iter.go#L902)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

don't forget to remove this before it merges
</blockquote></details>

Done.

**Pull request merged**
